### PR TITLE
Reorganized the reflection code to improve re-use and clarity.

### PR DESCRIPTION
**Pull request by tfenne**
I've left this as three commits for now, in case it's helpful to view the first two separately.  I'm pretty happy with where this ended up and the fact that it was minimal effort to modify the tests!

### CODE REVIEW

File: core/src/main/scala/dagr/core/cmdline/parsing/ClpReflectiveBuilder.scala [hunk 1]
```diff
@@ -22,25 +22,110 @@
...
+    }
+    else {
+      val argTypeSymbol = mirror.typeOf[ArgAnnotation]
+      val clpCtorSymbol = mirror.typeOf[CLPConstructorAnnotation]
+
+      // Check to see if we can find a constructor with @Arg annotations on it's parameters, or @CLPConstructor on it
+      val argOption = constructors.find(c => c.asMethod.paramLists.head.exists(p => p.annotations.exists(a => a.tree.tpe == argTypeSymbol)))
+      val clpOption = constructors.find(p => p.annotations.exists(a => a.tree.tpe == clpCtorSymbol))
+
+      argOption.getOrElse(clpOption.getOrElse(throw new IllegalStateException("FML"))).asMethod
```
Comment (nh13) [inline core/src/main/scala/dagr/core/cmdline/parsing/ClpReflectiveBuilder.scala:64]: so true.

File: core/src/main/scala/dagr/core/cmdline/parsing/ClpReflectiveBuilder.scala [hunk 2]
```diff
@@ -22,25 +22,110 @@
...
+                                       constructableType: Class[_],
+                                       typeName: String): ArgumentDescription = {
+    val jParam = jConstructor.getParameters()(index)
+    val annotation = Option(jParam.getAnnotation(classOf[ArgAnnotation]))
+    new ArgumentDescription(declaringClass, index, name, defaultValue, argumentType, unitType, constructableType, typeName, annotation)
+  }
+}
+
+/**
+  * Extension to the default ArgumentLookup to support lots more names that come from the
```
Comment (nh13) [inline core/src/main/scala/dagr/core/cmdline/parsing/ClpReflectiveBuilder.scala]: lots -> a lot

File: core/src/main/scala/dagr/core/cmdline/parsing/DagrCommandLineParser.scala [hunk 3]
```diff
@@ -229,29 +229,29 @@ class DagrCommandLineParser(val commandLineName: String, val includeClassesOmitt
...
+              pipelineGroup = Some(clp.pipelineGroup.newInstance)
+            }
+            catch {
+              case e: InstantiationException => throw new RuntimeException(e)
+              case e: IllegalAccessException => throw new RuntimeException(e)
+            }
+            taskGroupClassToTaskGroupInstance.put(clp.pipelineGroup, pipelineGroup.get)
+          }
+          var pipelines: ListBuffer[PipelineClass] = tasksByGroup.get(pipelineGroup.get)
+          if (null == pipelines) {
```
Comment (nh13) [inline core/src/main/scala/dagr/core/cmdline/parsing/DagrCommandLineParser.scala:248]: it's crazy that scala does not have a mutable `TreeMap`.

File: core/src/main/scala/dagr/core/cmdline/parsing/ParsingUtil.scala [hunk 4]
```diff
@@ -124,7 +95,7 @@ private[parsing] object ParsingUtil {
     }
 
     // Finally, make the map
-    classes.map(clazz => Tuple2(clazz, getAnnotation(clazz, classOf[CLPAnnotation]))).toMap
+    classes.map(clazz => Tuple2(clazz, ReflectionUtil.findJavaAnnotation(clazz, classOf[CLPAnnotation]).get)).toMap
```
Comment (nh13) [inline core/src/main/scala/dagr/core/cmdline/parsing/ParsingUtil.scala:98]: why not just put an `import ReflectionUtil._` inside this class so we don't have to write out `ReflectionUtil` each time

File: core/src/main/scala/dagr/core/util/ReflectionUtil.scala [hunk 5]
```diff
@@ -99,4 +128,112 @@ object ReflectionUtil {
...
+      tree.productElement(1).asInstanceOf[List[Tree]].map(p => reifyAnnotationParameter(p)).toArray[Any]
+    case other =>
+      throw new RuntimeException("Don't know how to handle a: " + other)
+  }
+
+  /** Takes a TermSymbol representing a java Enum value, and returns the value. */
+  private def termToEnum(term: TermSymbol): Any = {
+    val enumClass =  if (term.owner.isModuleClass) {
+      // This implies that the enum is an inner class, so we have to fetch it's class this weird way
+      typeToClass(mirror.reflectClass(term.owner.companion.asClass).symbol.toType)
```
Comment (nh13) [inline core/src/main/scala/dagr/core/util/ReflectionUtil.scala:211]: TIL

File: core/src/main/scala/dagr/core/util/ReflectiveBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+                              typeName : String) : ArgumentDescription = {
+    new Argument(
+      declaringClass=clazz, index=index, name=name, defaultValue=defaultValue,
+      argumentType = argumentType, unitType=unitType, constructableType = constructableType, typeName=typeName
+    ).asInstanceOf[ArgumentDescription]
+  }
+
+  /** Constructs a new instance of type T with the parameter values set on the builder (including defaults). */
+  def build() : T = {
+    val xs = argumentLookup.ordered.filter(arg => !arg.hasValue).map(arg => arg.name)
```
Comment (nh13) [inline core/src/main/scala/dagr/core/util/ReflectiveBuilder.scala]: use `filterNot(_.hasValue)`

File: core/src/main/scala/dagr/core/cmdline/parsing/ParsingUtil.scala (hunk 4 above)
Comment (tfenne) [inline core/src/main/scala/dagr/core/cmdline/parsing/ParsingUtil.scala:98]: It's personal preference.  So long as it's not super long, I prefer to see the object name at the usage site, to give me a clue where the symbol is coming from without having to check.

**Comment by nh13** [comment]
@tfenne, seriously, this must have hurt your brain.  Feel free to merge when ready.  To me, it's just:

magic

**Pull request merged**